### PR TITLE
Assumptions in `XStream2EncodingTest` are not correctly implemented

### PR DESCRIPTION
**Pull request by basil**
The assumptions in `XStream2EncodingTest`, added in commit b14a7a36e4, were designed to skip the test if the reflection-based trick to modify the default encoding failed to take effect. However, because the assumptions are in `@Before`/`@After` methods rather than `@Test` methods, they never worked, but we never noticed because when running in Java < 17 the assumption was always true. Starting in Java 17, the assumption is now false, but the test is not skipped as expected. Rather we get

<details>
<summary>Stack trace</summary>
<br>
<pre>
[ERROR] toXMLUnspecifiedEncoding  Time elapsed: 0.001 s  <<< FAILURE!
org.opentest4j.MultipleFailuresError: 
Multiple Failures (2 failures)
        org.junit.TestCouldNotBeSkippedException: Test could not be skipped due to other failures
        org.junit.TestCouldNotBeSkippedException: Test could not be skipped due to other failures
        at org.junit.vintage.engine.execution.TestRun.getStoredResultOrSuccessful(TestRun.java:196)
        at org.junit.vintage.engine.execution.RunListenerAdapter.fireExecutionFinished(RunListenerAdapter.java:226)
        at org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:192)
        at org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:79)
        at org.junit.runner.notification.SynchronizedRunListener.testFinished(SynchronizedRunListener.java:87)
        at org.junit.runner.notification.RunNotifier$9.notifyListener(RunNotifier.java:225)
        at org.junit.runner.notification.RunNotifier$SafeNotifier.run(RunNotifier.java:72)
        at org.junit.runner.notification.RunNotifier.fireTestFinished(RunNotifier.java:222)
        at org.junit.internal.runners.model.EachTestNotifier.fireTestFinished(EachTestNotifier.java:38)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:372)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
        at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
        at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
        at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
        at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:150)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
        Suppressed: org.junit.TestCouldNotBeSkippedException: Test could not be skipped due to other failures
                at org.junit.runners.model.MultipleFailureException.<init>(MultipleFailureException.java:36)
                at org.junit.internal.runners.model.MultipleFailureException.<init>(MultipleFailureException.java:10)
                at org.junit.runners.model.MultipleFailureException.assertEmpty(MultipleFailureException.java:102)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:39)
                at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
                at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
                at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
                ... 25 more
        Caused by: org.junit.AssumptionViolatedException: got: <java.lang.reflect.InaccessibleObjectException: Unable to make field private static volatile java.nio.charset.Charset java.nio.charset.Charset.defaultCharset accessible: module java.base does not "opens java.nio.charset" to unnamed module @887af79>, expected: null
                at org.junit.Assume.assumeThat(Assume.java:106)
                at org.junit.Assume.assumeNoException(Assume.java:153)
                at hudson.util.XStream2EncodingTest.clearDefaultEncoding(XStream2EncodingTest.java:63)
                at hudson.util.XStream2EncodingTest.useNonUTF8(XStream2EncodingTest.java:48)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:568)
                at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
                at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
                at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
                at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
                at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
                ... 28 more
        Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private static volatile java.nio.charset.Charset java.nio.charset.Charset.defaultCharset accessible: module java.base does not "opens java.nio.charset" to unnamed module @887af79
                at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
                at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
                at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
                at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
                at hudson.util.XStream2EncodingTest.clearDefaultEncoding(XStream2EncodingTest.java:60)
                ... 39 more
        Suppressed: org.junit.TestCouldNotBeSkippedException: Test could not be skipped due to other failures
                at org.junit.runners.model.MultipleFailureException.<init>(MultipleFailureException.java:36)
                at org.junit.internal.runners.model.MultipleFailureException.<init>(MultipleFailureException.java:10)
                at org.junit.runners.model.MultipleFailureException.assertEmpty(MultipleFailureException.java:102)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:39)
                at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
                at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
                at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
                ... 25 more
        Caused by: org.junit.AssumptionViolatedException: got: <java.lang.reflect.InaccessibleObjectException: Unable to make field private static volatile java.nio.charset.Charset java.nio.charset.Charset.defaultCharset accessible: module java.base does not "opens java.nio.charset" to unnamed module @887af79>, expected: null
                at org.junit.Assume.assumeThat(Assume.java:106)
                at org.junit.Assume.assumeNoException(Assume.java:153)
                at hudson.util.XStream2EncodingTest.clearDefaultEncoding(XStream2EncodingTest.java:63)
                at hudson.util.XStream2EncodingTest.clearDefaultEncodingAfter(XStream2EncodingTest.java:54)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:568)
                at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
                at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
                at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
                at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
                ... 28 more
        Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private static volatile java.nio.charset.Charset java.nio.charset.Charset.defaultCharset accessible: module java.base does not "opens java.nio.charset" to unnamed module @887af79
                at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
                at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
                at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
                at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
                at hudson.util.XStream2EncodingTest.clearDefaultEncoding(XStream2EncodingTest.java:60)
                ... 38 more
</pre>
</details>

which this PR resolves by moving the assumptions to `@Test` methods instead. So now the assumptions take effect as originally intended and the test is skipped when the assumption is false (i.e., on Java 17+).

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
